src/devices/MettlerToledo/MettlerToledoDevice.cs

```diff
@@ -0,0 +1,362 @@
+                var line = timeout == null ? _serialPort.ReadLine() : ReadSerialPortWithTimeout(start - DateTime.Now + timeout.Value);
+                var split = line.Split(' ', StringSplitOptions.RemoveEmptyEntries);
+
+                switch (split[0])
+                {
+                    case Errors.SYNTAX_ERROR:
+                        throw new SyntaxErrorException(command);
+                    case Errors.TRANSMISSION_ERROR:
+                        throw new TransmissionErrorException(command);
+                    case Errors.LOGICAL_ERROR:
+                        throw new LogicalErrorException(command);
+                }
+
+                if (split[0] == responseCode)
+                {
+                    return split;
+                }
+                else
+                {
+                    // if we're listening for weight change events, we will receive "S" responses periodically
+                    if (split[0] == Responses.WEIGHT_UPDATE)
+                    {
+                        WeightUpdated?.Invoke(this, ParseWeightReading(split));
+                    }
+                    else
+                    {
+                        throw new InvalidDataReceivedException(command, line);
+                    }
+                }
+            }
+
+            throw new TimeoutException();
+        }
+
+        private string ReadSerialPortWithTimeout(TimeSpan timeout)
+        {
+            if (timeout.TotalMilliseconds > _serialPort.ReadTimeout)
+            {
+                timeout = TimeSpan.FromMilliseconds(_serialPort.ReadTimeout);
+            }
+
+            var task = Task.Run(() => _serialPort.ReadLine());
+            if (task.Wait(timeout))
+            {
+                return task.Result;
+            }
+
+            throw new TimeoutException();
+        }
+
+        private MettlerToledoWeightReading GetWeightReading(bool forceImmediate = false)
+        {
+            var command = forceImmediate ? Commands.SEND_IMMEDIATE_WEIGHT_VALUE : Commands.SEND_STABLE_WEIGHT_VALUE;
+            var response = SendCommandWithResponse(command, Responses.WEIGHT_UPDATE);
+
+            // Check if we have an error
+            switch (response[1])
+            {
+                case Responses.COMMAND_NOT_EXECUTABLE:
+                    throw new CommandNotExecutableException(command);
+                case Responses.Range.UPPER_RANGE_EXCEEDED:
+                    throw new BalanceOutOfRangeException(true, false, command);
+                case Responses.Range.LOWER_RANGE_EXCEEDED:
+                    throw new BalanceOutOfRangeException(false, true, command);
+            }
+
+            return ParseWeightReading(response);
+        }
+
+        private MettlerToledoWeightReading ParseWeightReading(string[] response)
+        {
+            var value = double.Parse(response[2]);
+            var unit = _unitMapping.First(x => x.mtAbbreviation == response[3]).unitsNetUnit;
```

Comment:
Even if I love Linq, I would avoid using it and just browse the collection the normal way. Reason is because we're working on porting automatically bindings to the nanoFramework world to have those kind of bindings working on tiny devices such as ESP32. And there generics and Linq are not supported.